core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java

```diff
@@ -138,11 +138,10 @@ protected CqlRequestHandler(
     this.startTimeNanos = System.nanoTime();
     this.logPrefix = sessionLogPrefix + "|" + this.hashCode();
     LOG.trace("[{}] Creating new handler for request {}", logPrefix, statement);
-
-    this.initialStatement = statement;
     this.session = session;
     this.keyspace = session.getKeyspace().orElse(null);
     this.context = context;
+    this.initialStatement = Conversions.resolveStatement(statement, context);
```

Comment:
I agree that this is _one_ way to address the problem: before you send a batch statement make sure that you've got a batch type.  But this has a couple of very unpleasant consequences:

* You _have_ to introduce something like an "unknown" batch type to represent the state of batches which haven't had anything specified yet because they're going to use the default at execution time.
* In driver 4.x statements are immutable, so every time you have to use the default batch type you're creating a copy of the input BatchStatement object with exactly one field changed.

The much simpler answer is to enforce the presence of a batch type _at creation time_.  The builder interface at BatchStatement.builder() already requires a non-null BatchType so all we need to worry about are the BatchStatement.newInstance() types.  In that case it seems like we can make a modified version which accepts a non-null DriverContext as a param instead of a BatchType.  We can then use that DriverContext to do a lookup for the configured default batch type.  So long as our config defines a default (which it certainly should) we'll always generate a BatchStatement with a non-null batch type using that method.

Reply:
Sorry for Not familiar with the codebase , and I have got this one question here ;do you mean  something like : BatchStatement.newInstance(DriverContext context)? It seems  DriverContext can ony be get from the variable of  session, so what kind of usage of the interface can user  use ? That is what  I am asked at the beigining , is https://github.com/Maxwell-Guo/java-driver/blob/4.x-JAVA-3030/integration-tests/src/test/java/com/datastax/oss/driver/core/cql/BatchStatementUseConfigIT.java#L82  people need ? or something eles.

Reply:
Yes, BatchStatement.newInstance(DriverContext ctx) is exactly what I had in mind.  Keep in mind that the user will have to establish a Session of some kind in order to execute any of these statements.  The ["Quick start" example](https://github.com/datastax/java-driver/tree/4.x/manual/core#quick-start) in the manual demonstrates a really simple example.  For batch statements the user would be creating a BatchStatement object and then adding other statement objects to it but you can see how the general API works here.

The upshot is that being able to extract a DriverContext from the Session shouldn't be a problem when the user goes about creating a BatchStatement instance.